reducers/plugins/index.js

```diff
@@ -137,9 +170,16 @@ function components(state = {}, action) {
         if (action.name && action.data) {
             const nextState = {...state};
             const currentArray = nextState[action.name] || [];
-            const nextArray = [...currentArray];
+            if (action.name === 'PostDropdownMenu' && action.data.parentMenuId) {
+                const subMenuData = buildSubMenu(currentArray, action.data);
+                if (!subMenuData) {
+                    return state;
+                }
+                action.data = subMenuData;
+            }
+            const nextArray = [...currentArray.filter((c) => !c.subMenu || c.pluginId !== action.data.pluginId), action.data];
```

Comment:
i was hoping that we wouldn't have to filter anything out if we already had cleaned up things prior to getting here, would that be possible? Also `!c.subMenu` should never be `true` because of https://github.com/mattermost/mattermost-webapp/pull/3357/files#diff-b62c6fcf6082fdbefb2139d0ea16d5a1R176, no? same goes for `c.pluginId !== action.data.pluginId`, no?

Reply:
Sounds good.  I added the filter of currentArray to its creation.  `!c.subMenu` does toggle in values, however you are correct `c.pluginId !== action.data.pluginId` is always false.  I rearranged some of the logic based on your recommendations.